WIP Sinatra migration

AriettaZ opened the pull request:
The main page is almost done and the icones on the nav bar will be added later. The progess bar will push the number to the right side only if the width is smaller than 10%(may be more readable than all the numbers on the right?) I'm working on the subpages of different clusters. 
![image](https://user-images.githubusercontent.com/22674713/41561894-18e93598-7319-11e8-81cd-8650073dc2c6.png)

ericfranz commented:
1. In regards to the progress bar:

    >The progress bar will push the number to the right side only if the width is smaller than 10%(may be more readable than all the numbers on the right?)

    I'd keep the numbers to the right side for consistency. It will simplify the code a little bit further too.
2. For your commits, add more detailed commit messages. See https://robots.thoughtbot.com/5-useful-tips-for-a-better-commit-message for some helpful tips. For the existing commits, you can actually do an interactive rebase to replay the commits and quickly edit all of the commit messages with something more appropriate. Remind me and I'll show you how to do this next week when I'm back in the office.
3. Rename the "classes" directory to "lib"
4. Inline the `public/scripts/application.js` and `public/stylesheet/application.css` into the layout in `<style>` tag in the head and `<script>` tag near the end of the page. There isn't much custom css/js in those files and inlining will solve the cache busting problem
5. I don't think you need config/configuration_singleton.rb 
6. See https://github.com/OSC/ood-fileeditor/pull/80#issuecomment-398509939 on adding bin/setup
7. remove versioning `vendor/bundle` by taking 2 steps in 2 separate commits: 1. `git rm -r vendor/bundle` and 2. update git ignore to hide vendor/bundle (this should re-appear after running bin/setup)

ericfranz commented on views/layout.erb:
```diff
@@ -68,7 +126,26 @@
+            i++;
+          }
+          setTimeout(refreshImages, 5000);
+        })();
```
I know I said to move this to the layout, but this looks like it is specific to the ganglia graphs, so maybe this block should just be at the bottom of the ganglia graphs view.

Also, I think this block can be refactored and simplified. The "refreshImages" function is essentially

1. "get all of the updatable elements that are visible", and 
2. call "startRefresh" with these as an argument

The fix:

1. For "get all of the updatable elements that are visible", `$('.updateable:visible')` should work. No need for a while loop.
2. The array [forEach](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach) passes the value as the first argument, so this might work too: `$('.updateable:visible').toArray().forEach(startRefresh)`

At this point the code might be easier to read if the body of `startRefresh` be in an anonymous function instead of a function expression (setting function to variable). Then you could use jQuery's built in each method (I believe `this` is the value in that case) i.e. `$('.updateable:visible').each(function(){ ... });`

ericfranz commented on views/_system_large.erb:
```diff
@@ -6,7 +6,7 @@
             <div class="panel-heading"><h4 class="panel-title">Nodes</h4></div>
             <div class="panel-body">
-              <img class="img-responsive center-block updateable" alt="Graph" src="<%= ganglia.small.report_moab_nodes.png %>" onclick=" $('<%= #large-image#{@count} %>').replaceWith(
+              <img class="img-responsive center-block updateable" alt="Graph" src="<%= ganglia.small.report_moab_nodes.png %>" onclick=" $('#large-image<%= @count %>').replaceWith(
```
FYI the reason that this doesn't work:

```erb
onclick=" $('<%= #large-image#{@count} %>')
```

is that the ruby expression is missing quotes:

```erb
onclick=" $('<%= "#large-image#{@count}" %>')
```

AriettaZ commented:
I made some changes to the code and all the pages work functionally now. 
1. The progress bar is changed back to data on the right side
2. I updated <script> and it's still in the layout.erb for now. I'll move it to system_large.erb and system_status.erb later. The .setSearch is a method in URI.js, I received "Uncaught TypeError: (intermediate value).setSearch is not a function" and is trying to solve.
current js:
```
<script>
    $(document).ready(function(){
        $('[data-toggle="tooltip"]').tooltip();
        setTimeout($(".updateable:visible").each(function(){
           return this.src = new URL(this.src).setSearch("timestamp", (new Date).getTime()).toString();
        }),5000);
    });
</script>
```
3. The "classes" directory were renamed to "lib"
4. config/configuration_singleton.rb is moved
5. I removed vendor/bundle and tested new bin/setup, it works

ericfranz commented:
As discussed:

- [ ] add jquery you are using to public/ and update layout to use that
- [ ] add font awesome you are using to public/ and update layout to use that
- [ ] update error pages to use simple erb
- [ ] remove unnecessary configuration
- [ ] remove .env files
- [ ] remove about route
- [ ] redirect from get "/" to get "/clusters"
- [ ] use `ENV['OOD_DASHBOARD_TITLE'] || "Open OnDemand"` for title
- [ ] fix font awesome issue so icons load like before

Also, just noticed this, but instead of parsing the cluster config on every request, it would probably be better to parse this a single time at application start. You could do this at load time and assign the clusters list to a constant, like `CLUSTERS`.

Also

```ruby
get '/clusters/:id' do
  id=params[:id].to_sym
  cluster = @oodclusters[id]|| nil
```

The `|| nil` is redundant as `@oodclusters[id]` will return nil if id is invalid.

AriettaZ commented:
I've updated and cleaned up the project and have font awesome issue fixed. It's `icon('area-chart')` for rails so I was using `<i class="fa fa-area-chart"></i>` which doesn't work. The fix is changing it into `<i class="fa fa-chart-area"></i>`.  

One issue I have is to load the clusters at application start and assign it to a constant. I have `@clusters||=valid_clusters `as following for now. Should I change the `@clusters` into a global constant like `@CLUSTERS` or an instance variable like `CLUSTERS` and pass it to other files? 
```
before do
  @clusters||=valid_clusters
end
```

To load it when starting the application, should I make the app.rb a class inherited from  Sinatra::Application and get the `@clusters` in the constructor like the following? If so, I'm not sure can I use the helper method "valid_clusters" in the constructor. I may need to create a helper module or put valid_clusters code to the constructor. 
```
class SystemstatusApp <  Sinatra::Application
    def initialize(app=nil)
        super()
        @clusters=valid_clusters
    end
    ...
    ...
end
```

Other than this, I'll add more comments. Overall, the systemstatus app works and I'm happy with it.

ericfranz commented:
I think the two options are:

#### Option 1

Switch to modular style (i.e. subclass `Sinatra::Application`) then make the filtered clusters list a class variable i.e.

```ruby
class SystemStatusApp <  Sinatra::Application
  class << self
    def clusters
      # return @clusters if defined

      # use begin/rescue to set @clusters - see below for begin/rescue block except set @clusters
    end
  end
end
```

Then SystemStatusApp.clusters is the OodCore::Clusters object. Here is an example of this memoization technique: https://github.com/OSC/ood-dashboard/blob/7622025dc28f59d71fbc94b0e0949bf4653c6c90/config/configuration_singleton.rb#L84-L88

#### Option 2

Or define a constant:

```ruby
require 'sinatra'
require 'ood_core'
require 'moab'

Dir[File.dirname(__FILE__) + "/lib/*.rb"].each {|file| require_relative file }

CLUSTERS = (ENV['OOD_CLUSTERS'] || '/etc/ood/config/clusters.d').tap do |config|
  begin
    OodCore::Clusters.new(OodCore::Clusters.load_file(config).select(&:job_allow?)
          .select { |c| c.custom_config[:moab] }
          .select { |c| c.custom_config[:ganglia] }
          .reject { |c| c.metadata.hidden }
    )
  rescue OodCore::ConfigurationNotFound
    OodCore::Clusters.new([])
  end
end
```


The question is what is more readable code? I would do the constant first, once that is complete, make a separate branch (off of the sinatra branch) and convert app.rb to "modular style" to compare the two.

AriettaZ commented:
I changed the `@clusters` to `CLUSTERS` by taking the option 2 and the site remains the same functionality. 
```
require 'sinatra'
require 'ood_core'
require 'moab'

Dir[File.dirname(__FILE__) + "/lib/*.rb"].each {|file| require_relative file }

begin
  CLUSTERS = OodCore::Clusters.new(OodCore::Clusters.load_file(ENV['OOD_CLUSTERS'] || '/etc/ood/config/clusters.d').select(&:job_allow?)
            .select { |c| c.custom_config[:moab] }
            .select { |c| c.custom_config[:ganglia] }
            .reject { |c| c.metadata.hidden }
          )
rescue OodCore::ConfigurationNotFound
  CLUSTERS = OodCore::Clusters.new([])
end
```

ericfranz commented on views/index.erb:
```diff
@@ -1,6 +1,6 @@
-    <% if index == @clusters.count - 1 && @clusters.count.odd? %>
+<% CLUSTERS.each_with_index do |cluster, index | %>
+    <% if index == CLUSTERS.count - 1 && CLUSTERS.count.odd? %>
         <div class="col-sm-3"></div>
```
Instead of adding an empty div I think you can instead add an offset class to `<div class="col-sm-6 col-xs-12">` https://getbootstrap.com/docs/3.3/css/#grid-offsetting i.e.

```ruby
<div class="col-sm-offset-3 col-sm-6 col-xs-12">
```

Which means you could do something like:

```ruby
<div class="<%= 'col-sm-offset-3' if index == CLUSTERS.count - 1 && CLUSTERS.count.odd? %> col-sm-6 col-xs-12">
```

removing the control flow.

Or you could just get rid of this shift, I'm not sure what it adds. Centering the bottom cluster seems to draw more attention to it over the others. Without the shift it looks like this:

![screen 2018-06-25 at 5 45 42 pm](https://user-images.githubusercontent.com/512333/41877548-b075e362-789f-11e8-896d-5980ea299a31.png)

ericfranz replied:
Keep the centering. Alan likes it.

ericfranz commented:
@achalker do you care at all whether the system status app looks like

![](https://user-images.githubusercontent.com/512333/41877548-b075e362-789f-11e8-896d-5980ea299a31.png)

or

![screen 2018-06-25 at 5 56 54 pm](https://user-images.githubusercontent.com/512333/41877943-2bd5730a-78a1-11e8-9ce7-852e225bfa98.png)

I was just suggesting possibly removing the extra code that centers the bottom cluster.

achalker commented:
I do like the centered look better, but if there is a good reason for removing it I’m fine with doing that.


---------------------------
Alan Chalker, Ph.D.
alanc@osc.edu<mailto:alanc@osc.edu>
614-247-8672

From: Eric Franz <notifications@github.com>
Sent: Monday, June 25, 2018 5:58 PM
To: AweSim-OSC/osc-systemstatus <osc-systemstatus@noreply.github.com>
Cc: achalker <github@chalker.org>; Mention <mention@noreply.github.com>
Subject: Re: [AweSim-OSC/osc-systemstatus] WIP Sinatra migration (#53)


@achalker<https://github.com/achalker> do you care at all whether the system status app looks like

[https://user-images.githubusercontent.com/512333/41877548-b075e362-789f-11e8-896d-5980ea299a31.png]<https://user-images.githubusercontent.com/512333/41877548-b075e362-789f-11e8-896d-5980ea299a31.png>

or

[screen 2018-06-25 at 5 56 54 pm]<https://user-images.githubusercontent.com/512333/41877943-2bd5730a-78a1-11e8-9ce7-852e225bfa98.png>

I was just suggesting possibly removing the extra code that centers the bottom cluster.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/AweSim-OSC/osc-systemstatus/pull/53#issuecomment-400109161>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AGe2ctCITF_Im5p1IZGk73Kygj5qi2bBks5uAVzZgaJpZM4UsemK>.

ericfranz commented:
- [ ] Update centering code https://github.com/AweSim-OSC/osc-systemstatus/pull/53#pullrequestreview-131796791 to use offset instead of empty dive
- [ ] Rename not_found.erb to 404.erb
- [ ] Address the empty Rakefile. Will you be removing it or adding the test setup like in https://github.com/OSC/ood-example-ps/blob/master/Rakefile and the test folder? Up to you.

ericfranz commented:
Oh I see. Two approaches:

https://github.com/AweSim-OSC/osc-systemstatus/blob/c670ee40c5c26c8b5bea966a36063d049386d583/app.rb#L57-L65

and

https://github.com/OSC/ood-fileeditor/blob/600b3b64de1576ea41bfc8c2eb7e3cb5187d3bdf/app.rb#L101-L111 (note: I did suggest to Tyler to rename those erbs from `404.html.erb` to `404.erb` so then it would be `erb :404`

And then there are multiple ways to raise a 404:

```
raise Sinatra::NotFound
halt 404
```

I actually don't know what the preferred style should be....will have to think about this.

ericfranz commented:
Probably I'd change the views to 404.erb and 500.erb since when looking at the list of views these stand out, and its similar to the Rails convention.

In regards to `raise Sinatra::NotFound` vs `halt(404)`. `halt` uses `throw`. It might be an interesting read to see the difference between `halt` and `raise` in Ruby. Seems like here both work so it probably doesn't matter.

AriettaZ commented:
I deleted Rakefile and updated index.erb. To use 404.erb and 500.erb, I used `erb :'404'` and  `erb :'500'` instead of `erb :404` and  `erb :500` because they will give me a syntax error as the following. I don't know the reason yet but   `erb :'404'` and  `erb :'500'` work fine. 


```
/users/PZS0002/azhu/ondemand/dev/systemstatus/app.rb:59: syntax error, unexpected tINTEGER, expecting tSTRING_CONTENT or tSTRING_DBEG or tSTRING_DVAR or tSTRING_END
  erb :404
          ^ (SyntaxError)
  config.ru:1:in `require_relative'
  config.ru:1:in `block in <main>'
  /users/PZS0002/azhu/ondemand/dev/systemstatus/vendor/bundle/ruby/gems/rack-2.0.5/lib/rack/builder.rb:55:in `instance_eval'
  /users/PZS0002/azhu/ondemand/dev/systemstatus/vendor/bundle/ruby/gems/rack-2.0.5/lib/rack/builder.rb:55:in `initialize'
  config.ru:1:in `new'
  config.ru:1:in `<main>'
  /opt/rh/rh-passenger40/root/usr/share/passenger/helper-scripts/rack-loader.rb:108:in `eval'
  /opt/rh/rh-passenger40/root/usr/share/passenger/helper-scripts/rack-loader.rb:108:in `load_app'
  /opt/rh/rh-passenger40/root/usr/share/passenger/helper-scripts/rack-loader.rb:127:in `<module:App>'
  /opt/rh/rh-passenger40/root/usr/share/passenger/helper-scripts/rack-loader.rb:27:in `<module:PhusionPassenger>'
  /opt/rh/rh-passenger40/root/usr/share/passenger/helper-scripts/rack-loader.rb:26:in `<main>'
Error ID
```

ericfranz commented:
Use this for favicon and remove embedded favicon

```
"#{ENV['OOD_PUBLIC_URL'] || '/public'}/favicon.ico"

# or

"#{public_url}/favicon.ico"
```

ericfranz commented on app.rb:
```diff
@@ -0,0 +1,68 @@
+    "/pun/sys/dashboard/"
+  end
+
+  def relative_url
```
This method should not be needed; use Sinatra's `url(` helper for app url's in the views. For example, url('/') should take you to the root of the app.

ericfranz commented on views/layout.erb:
```diff
@@ -0,0 +1,150 @@
+
+        <ul class="navbar-breadcrumbs">
+          <li><a href='<%=dashboard_url%>'><%= dashboard_title %></a></li>
+          <li><a href='<%=relative_url%>'>System Status</a></li>
```
instead, try `url("/")`

ericfranz commented:
I added two comments regarding the `relative_url` helper method. Once those are addressed I'll be ready to merge this PR.

ericfranz commented on .gitignore:
```diff
@@ -24,3 +24,4 @@
 /public/.htaccess
 /.env.local*
 /vendor/bundle
+/Gemfile.lock
```
@AriettaZ was this intentional? Did we have a discussion about this?

AriettaZ replied:
We didn't discuss this. I think we should add Gemfile.lock back to the repository because I added this line and forget to delete it.
